src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java

```diff
@@ -258,13 +267,22 @@ else if (length == 0) {
             return 0;
         }
 
-        int mark = offset;
-        while (length > 0) {
-            buf[offset++] = (byte) read();
-            length--;
+        if (stateLock.tryLock()) {
+            try {
+                int mark = offset;
+                while (length > 0) {
+                    buf[offset++] = (byte) read();
+                    length--;
+                }
+
+                return offset - mark;
+            }
+            finally {
+                stateLock.unlock();
+            }
         }
 
-        return offset - mark;
+        return 0;
```

Comment:
You cannot return 0 because the lock is not acquired. That's against the semantics of read() which should block until there is data or EOF.